scipy/signal/_polyutils.py

```diff
@@ -0,0 +1,171 @@
+
+To distinguish the two sets, the "new-style" routine names start with `npp_`
+"""
+import scipy._lib.array_api_extra as xpx
+
+
+def _sort_cmplx(arr, xp):
+    # xp.sort is undefined for complex dtypes. Here we only need some
+    # consistent way to sort a complex array, including equal magnitude elements.
+    arr = xp.asarray(arr)
+    if xp.isdtype(arr.dtype, 'complex floating'):
+        sorter = abs(arr) + xp.real(arr) + xp.imag(arr)**3
+    else:
+        sorter = arr
+
+    idxs = xp.argsort(sorter)
+    return arr[idxs]
+
+
+def polyroots(coef, *, xp):
+    """numpy.roots, best-effor replacement
+    """
+    if coef.shape[0] < 2:
+        return xp.asarray([], dtype=coef.dtype)
+
+    root_func = getattr(xp, 'roots', None)
+    if root_func:
+        # OK on numpy and jax, is broken on cupy
+        return root_func(coef)
+
+    # companion matrix
+    n = coef.shape[0]
+    a = xp.eye(n - 1, n - 1, k=-1, dtype=coef.dtype)
+    a[:, -1] = -xp.flip(coef[1:]) / coef[0]
+
+    # non-symmetric eigenvalue problem is not in the spec but is available on e.g. torch
+    if hasattr(xp.linalg, 'eigvals'):
+        return xp.linalg.eigvals(a)
+    else:
+        import numpy as np
+        return xp.asarray(np.linalg.eigvals(np.asarray(a)))
+
+
+# https://github.com/numpy/numpy/blob/v2.1.0/numpy/lib/_function_base_impl.py#L1874-L1925
+def _trim_zeros(filt, trim='fb'):
+    first = 0
+    trim = trim.upper()
+    if 'F' in trim:
+        for i in filt:
+            if i != 0.:
+                break
+            else:
+                first = first + 1
+    last = filt.shape[0]
+    if 'B' in trim:
+        for i in filt[::-1]:
+            if i != 0.:
+                break
+            else:
+                last = last - 1
+    return filt[first:last]
+
+
+# ### Old-style routines ###
+
+
+# https://github.com/numpy/numpy/blob/v2.2.0/numpy/lib/_polynomial_impl.py#L1232
+def _poly1d(c_or_r, *, xp):
+    """ Constructor of np.poly1d object from an array of coefficients (r=False)
+    """
+    c_or_r = xpx.atleast_nd(c_or_r, ndim=1, xp=xp)
+    if c_or_r.ndim > 1:
+        raise ValueError("Polynomial must be 1d only.")
+    c_or_r = _trim_zeros(c_or_r, trim='f')
+    if c_or_r.shape[0] == 0:
+        c_or_r = xp.asarray([0], dtype=c_or_r.dtype)
+    return c_or_r
+
+
+# https://github.com/numpy/numpy/blob/v2.2.0/numpy/lib/_polynomial_impl.py#L702-L779
+def polyval(p, x, *, xp):
+    """ Old-style polynomial, `np.polyval`
+    """
+    p = xp.asarray(p)
+    x = xp.asarray(x)
+    y = xp.zeros_like(x)
+
+    for pv in p:
+        y = y * x + pv
+    return y
+
+
+# https://github.com/numpy/numpy/blob/v2.2.0/numpy/lib/_polynomial_impl.py#L34-L157
+def poly(seq_of_zeros, *, xp):
+    # Only reproduce the 1D variant of np.poly
+    seq_of_zeros = xp.asarray(seq_of_zeros)
+    seq_of_zeros = xpx.atleast_nd(seq_of_zeros, ndim=1, xp=xp)
+
+    if seq_of_zeros.shape[0] == 0:
+        return 1.0
+
+    # prefer np.convolve etc, if available
```

Comment:
`np.convolve` might be an option to consider for `array-api-extra`. 

We cannot always default to `scipy.signal.convolve` or `scipy.ndimage.convolve` since those are CPU only.
CuPy and JAX have it available, and a there's a pytorch equivalent:
https://github.com/pytorch/pytorch/blob/main/torch/_numpy/_funcs_impl.py#L612

There is no usable pure-python, array-agnostic implementation though.